service/src/PlatformTopo.cpp

```diff
@@ -544,77 +635,90 @@ namespace geopm
         }
     }
 
-    FILE *PlatformTopoImp::open_lscpu(void)
+    bool PlatformTopoImp::check_file(const std::string &file_path)
     {
-        FILE *result = nullptr;
-        if (M_TEST_CACHE_FILE_NAME.size()) {
-            result = fopen(M_TEST_CACHE_FILE_NAME.c_str(), "r");
-            if (!result) {
-                throw Exception("PlatformTopoImp::open_lscpu(): Could not open test lscpu file",
-                                errno ? errno : GEOPM_ERROR_FILE_PARSE, __FILE__, __LINE__);
-            }
+        struct sysinfo si;
+        int err = sysinfo(&si);
+        if (err) {
+            throw Exception("PlatformTopoImp::check_file(): sysinfo err: ",
+                            errno ? errno : GEOPM_ERROR_RUNTIME, __FILE__, __LINE__);
+        }
+
+        struct stat file_stat;
+        err = stat(file_path.c_str(), &file_stat);
+        if (err) {
+            throw Exception("PlatformTopoImp::create_cache(): stat failure:",
+                            errno ? errno : GEOPM_ERROR_RUNTIME, __FILE__, __LINE__);
+        }
+
+        struct geopm_time_s current_time;
+        geopm_time_real(&current_time);
+
+        unsigned int last_boot_time = current_time.t.tv_sec - si.uptime;
+        if (file_stat.st_mtime < last_boot_time) {
+            return false; // file is older than last boot
         }
         else {
-            result = fopen(M_CACHE_FILE_NAME.c_str(), "r");
-            if (!result) {
-                int err = geopm_sched_popen("lscpu -x", &result);
-                if (err) {
-                    throw Exception("PlatformTopoImp::open_lscpu(): Could not popen lscpu command",
-                                    errno ? errno : GEOPM_ERROR_RUNTIME, __FILE__, __LINE__);
-                }
-                m_do_fclose = false;
+            mode_t expected_perms;
+            if (file_path.c_str() == M_SERVICE_CACHE_FILE_NAME) {
```

Comment:
This banks on the fact that when the service cache file name is passed in `file_path`, it will be passed as `M_SERVICE_CACHE_FILE_NAME` (which is how it should be, hence this const member's existence). Maybe I'm being too paranoid, but for full robustness, I would recommend using std::string's operator==.

If not for that, at least for consistency, seeing an earlier similar check is using std::string's operator==, as opposed to doing pointer comparison.

Reply:
Sorry, I'm way off. M_SERVICE_CACHE_FILE_NAME is of type std::string. If if it had been const char* my comment is incorrect.

The operator== that takes char* and std::string is being used. I guess the only thing worth noting is that calling c_str() is not necessary and both std::strings could be passed to operator==.